src/providers/ldap/sdap_child_helpers.c

```diff
@@ -263,6 +265,129 @@ static int parse_child_response(TALLOC_CTX *mem_ctx,
+    child->io->write_to_child_fd = -1;
+    talloc_set_destructor((TALLOC_CTX *)child->io, child_io_destructor);
+
+    *_child = child;
+
+    return EOK;
+}
+
+static errno_t parse_select_principal_response(TALLOC_CTX *mem_ctx,
+                                               uint8_t *buf, ssize_t size,
+                                               char **sasl_primary,
+                                               char **sasl_realm)
+{
+    uint32_t len = 0;
+    size_t p = 0;
+
+    SAFEALIGN_COPY_UINT32_CHECK(&len, buf + p, size, &p);
+    if (len > size - p) {
+        return EINVAL;
+    }
+    *sasl_primary = talloc_size(mem_ctx, sizeof(char) * (len + 1));
+    if (*sasl_primary == NULL) {
+        return ENOMEM;
+    }
+    safealign_memcpy(*sasl_primary, buf + p, sizeof(char) * len, &p);
+    (*sasl_primary)[len] = '\0';
+
+    SAFEALIGN_COPY_UINT32_CHECK(&len, buf + p, size, &p);
+    if (len > size - p) {
+        return EINVAL;
+    }
+    *sasl_realm = talloc_size(mem_ctx, sizeof(char) * (len + 1));
+    if (*sasl_realm == NULL) {
+        return ENOMEM;
+    }
+    safealign_memcpy(*sasl_realm, buf + p, sizeof(char) * len, &p);
+    (*sasl_realm)[len] = '\0';
+
+    DEBUG(SSSDBG_TRACE_LIBS, "result: '%s', '%s'\n", *sasl_primary, *sasl_realm);
+
+    return EOK;
+}
+
+errno_t sdap_select_principal_from_keytab_sync(TALLOC_CTX *mem_ctx,
+                                               const char *princ_str,
+                                               const char *realm_str,
+                                               const char *keytab_name,
+                                               char **sasl_primary,
+                                               char **sasl_realm)
+{
+    static uint8_t response[2048];
+    struct io_buffer *buf = NULL;
+    int ret;
+    struct sdap_child *child = NULL;
+    ssize_t len;
+
+    ret = alloc_child(mem_ctx, &child);
+    if (ret != EOK) {
+        return ret;
+    }
+
+    ret = create_child_req_send_buffer(mem_ctx, LDAP_CHILD_SELECT_PRINCIPAL,
+                                       realm_str, princ_str, keytab_name, 0,
+                                       &buf);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "create_child_req_send_buffer() failed.\n");
+        ret = EFAULT;
+        goto done;
+    }
+
+    ret = sdap_fork_child(NULL, child, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "sdap_fork_child() failed.\n");
+        goto done;
+    }
+
+    len = sss_atomic_write_s(child->io->write_to_child_fd, buf->data, buf->size);
+    if (len != buf->size) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "sss_atomic_write_s() failed\n");
+        ret = EIO;
+        goto done;
+    }
+
+    sdap_close_fd(&child->io->write_to_child_fd);
+
+    len = sss_atomic_read_s(child->io->read_from_child_fd,
+                            response, sizeof(response));
+    if (len <= 0) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "sss_atomic_read_s() failed\n");
```

Comment:
Hi,

in case there is an error like no principal found, keytab not readable or does not exist etc there will be just this generic error message. At least I would suggest to change the error message to something like "Failed to get principal from keytab, see ldap_child.log for details".  Ideally the PID would be mentioned as well to make searching ldap_child.log more easy but I'm not sure if it would be easy to get the PID here.

As an alternative `ldap_child` can return an error message, so if only a single item is returned it is an error message, if two items are returned they are sasl_primary and sasl_realm.

bye,
Sumit

Reply:
> I would suggest to change the error message to something like "Failed to get principal from keytab, see ldap_child.log for details". Ideally the PID would be mentioned as well

Done.